### ISSUE TITLE
Generator with unreachable yield not interpreted to be a generator

### ISSUE THREAD

**Issue by layday** [microsoft/pyright#1750]
**Describe the bug**
A generator function or async generator function with an unreachable yield (i.e. a generator which does not yield) is not interpreted to be a generator by Pyright.  This makes it difficult to type an overridable generator function which does not yield any values by default. For example:

```py
In [1]: def foo():
   ...:     yield
   ...:     return
   ...:

In [2]: [a for a in foo()]
Out[2]: [None]

In [3]: def foo():
   ...:     return
   ...:     yield
   ...:

In [4]: [a for a in foo()]
Out[4]: []

In [5]: def foo():
   ...:     if False:
   ...:         yield
   ...:

In [6]: [a for a in foo()]
Out[6]: []

```

**To Reproduce**
Type check the snippet below.

**Expected behavior**
That no errors would be produced.

**Screenshots or Code**

```python
# pyright: strict

async def foo():
    yield

async def bar():
    return
    yield

def baz():
    return
    yield

async def qux():
    [a async for a in foo()]

    # Type of "a" is unknown
    # "Coroutine[Any, Any, None]" is not iterable
    #   "__aiter__" method not defined
    [a async for a in bar()]

    # Type of "a" is unknown
    # "None" is not iterable
    [a for a in baz()]

```

**VS Code extension or command-line**
1.1.130 - works with 1.1.129.

**Comment by erictraut** [comment]
Thanks for the bug report. I agree that a function should still be considered a generator even if the yield statements are provably unreachable through static analysis. This will be fixed in the next release.

**Comment by erictraut** [comment]
This is fixed in pyright 1.1.131, which I just published. It was included in the previous version of pylance.
